src/test/regress/sql/aux_ao_rels_stat.sql

```diff
@@ -195,3 +195,174 @@ ORDER BY index;
 -- Drop AO and aux tables
 DROP TABLE ao_table;
 
+-- Test the fix: Adding data to a partitioned AO table using `COPY FROM`  
+-- does not change the `tupcount` in the segment table on the QD.
+CREATE OR REPLACE FUNCTION get_total_tupcount(table_name text)
+RETURNS bigint AS $$
+DECLARE
+    aoseg_table_name text;
+    tupcount_result bigint;
+BEGIN
+    -- Get the AO segment table name for the given partition
+    SELECT (n.nspname || '.' || c.relname) INTO aoseg_table_name
+    FROM pg_appendonly a
+        JOIN pg_class c ON c.oid = a.segrelid
+        JOIN pg_namespace n ON c.relnamespace = n.oid
+    WHERE a.relid = table_name::regclass::oid;
+
+    -- Sum tupcount from all segments
+    EXECUTE format('SELECT SUM(tupcount) FROM %s', aoseg_table_name)
+        INTO tupcount_result;
+
+    RETURN tupcount_result;
+END;
+$$ LANGUAGE plpgsql;
```

Comment:
```suggestion
CREATE OR REPLACE FUNCTION gp_ao_or_aocs_seg(rel regclass,
  segno OUT integer,
  tupcount OUT bigint,
  modcount OUT bigint,
  formatversion OUT smallint,
  state OUT smallint)
RETURNS SETOF record as $$
declare
  relstorage_var char;	/* in func */
begin	/* in func */
  select relstorage into relstorage_var from pg_class where oid = rel; /* in func */
  if relstorage_var = 'c' then	/* in func */
    for segno, tupcount, modcount, formatversion, state in SELECT DISTINCT x.segno, x.tupcount, x.modcount, x.formatversion, x.state FROM gp_toolkit.__gp_aocsseg(rel) x loop	/* in func */
      return next;	/* in func */
    end loop;	/* in func */
  else	/* in func */
    for segno, tupcount, modcount, formatversion, state in SELECT x.segno, x.tupcount, x.modcount, x.formatversion, x.state FROM gp_toolkit.__gp_aoseg(rel) x loop	/* in func */
      return next;	/* in func */
    end loop;	/* in func */
  end if;	/* in func */
end;	/* in func */
$$ LANGUAGE plpgsql;
```

Reply:
Could you please explain why you think it’s better to use the gp_toolkit.__gp_aoseg() function instead of directly reading from the auxiliary segment table? The patch modifies the tupcount in the segment table, and to me, the previous code seems simpler, as it reads only the data that the patch changes.

Reply:
I think it’s useful to show the direct state of the field in the table, including intermediate and top partition levels, to demonstrate that there are no segment tables at these levels. This helps provide a clearer understanding of how tupcount in segment tables is managed (or not) across different partition layers.

Reply:
This also works fine:
```sql
CREATE OR REPLACE FUNCTION get_total_tupcount(table_name text)
RETURNS bigint AS $$
DECLARE
    tupcount_result bigint;
    relstorage_var char;
BEGIN
    SELECT relstorage INTO relstorage_var FROM pg_class WHERE oid = table_name::regclass;

    IF relstorage_var = 'c' THEN
        -- Use gp_toolkit.__gp_aocsseg for column-oriented tables
        SELECT SUM(tupcount) INTO tupcount_result
        FROM (
            SELECT DISTINCT segno, tupcount, modcount, formatversion, state
            FROM gp_toolkit.__gp_aocsseg(table_name::regclass)
        ) subquery;
    ELSE
        -- Use gp_toolkit.__gp_aoseg for row-oriented tables
        SELECT SUM(tupcount) INTO tupcount_result
        FROM (
            SELECT DISTINCT segno, tupcount, modcount, formatversion, state
            FROM gp_toolkit.__gp_aoseg(table_name::regclass)
        ) subquery;
    END IF;

    RETURN tupcount_result;
END;
$$ LANGUAGE plpgsql;
```

Reply:
Okay, that makes it even easier.
```sql
CREATE OR REPLACE FUNCTION get_total_tupcount(table_name text) RETURNS bigint AS $$
BEGIN
    IF relstorage = 'c' FROM pg_class WHERE oid = table_name::regclass THEN
        -- Use gp_toolkit.__gp_aocsseg for column-oriented tables
        RETURN (SELECT SUM(DISTINCT tupcount)
            FROM gp_toolkit.__gp_aocsseg(table_name::regclass));
    ELSE
        -- Use gp_toolkit.__gp_aoseg for row-oriented tables
        RETURN (SELECT SUM(tupcount)
            FROM gp_toolkit.__gp_aoseg(table_name::regclass));
    END IF;
END;
$$ LANGUAGE plpgsql;
```

Reply:
additional optimizations
```suggestion
CREATE OR REPLACE FUNCTION get_total_tupcount(table_oid oid) RETURNS bigint AS $$
BEGIN
    IF relstorage = 'c' FROM pg_class WHERE oid = table_oid THEN
        -- Use gp_toolkit.__gp_aocsseg for column-oriented tables
        RETURN (SELECT SUM(DISTINCT tupcount)
            FROM gp_toolkit.__gp_aocsseg(table_oid));
    ELSE
        -- Use gp_toolkit.__gp_aoseg for row-oriented tables
        RETURN (SELECT SUM(tupcount)
            FROM gp_toolkit.__gp_aoseg(table_oid));
    END IF;
END;
$$ LANGUAGE plpgsql;
```

Reply:
Thanks for your additions, but I still can't understand why you think it's better to use a function that outputs information by columns. Rather than accessing the segment table directly? Are we sure that the number of tuples will always be the same in all columns regardless of the memory segmentation settings? And what's wrong with my approach?

Reply:
I did not find any tests in the project that use these functions. Meanwhile, direct reading from `pg_aocseg*` is used, for example, in the same file in other tests. Also, [according to the documentation](https://docs.vmware.com/en/VMware-Greenplum/6/greenplum-database/ref_guide-removed-objects.html?hWord=N4IghgNiBcIOYAcD6BXBATMAXApksA9kgLZgDOuATkhdmSAL5A), it seems that the upstream intended to remove these functions, but for some reason, they didn't.

Reply:
okay, how about such optimization
```suggestion
CREATE OR REPLACE FUNCTION get_total_tupcount(table_oid oid, out tupcount_result bigint)
RETURNS bigint AS $$
BEGIN
    EXECUTE format('SELECT SUM(tupcount) FROM pg_aoseg.%s', (
        SELECT c.relname FROM pg_appendonly a
        JOIN pg_class c ON c.oid = a.segrelid
        WHERE a.relid = table_oid
    )) INTO tupcount_result;
END;
$$ LANGUAGE plpgsql;
```